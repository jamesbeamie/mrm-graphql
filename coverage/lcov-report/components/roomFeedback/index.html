<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for components/roomFeedback</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> components/roomFeedback
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">0% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>0/406</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">0% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>0/136</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">0% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>0/42</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">0% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>0/388</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line low'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="AddQuestion.js"><a href="AddQuestion.js.html">AddQuestion.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="68" class="abs low">0/68</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="30" class="abs low">0/30</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="68" class="abs low">0/68</td>
	</tr>

<tr>
	<td class="file low" data-value="DeleteFeedback.js"><a href="DeleteFeedback.js.html">DeleteFeedback.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="26" class="abs low">0/26</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="26" class="abs low">0/26</td>
	</tr>

<tr>
	<td class="file low" data-value="EditFeedback.js"><a href="EditFeedback.js.html">EditFeedback.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="79" class="abs low">0/79</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="34" class="abs low">0/34</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="5" class="abs low">0/5</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="77" class="abs low">0/77</td>
	</tr>

<tr>
	<td class="file low" data-value="Feedback.js"><a href="Feedback.js.html">Feedback.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="14" class="abs low">0/14</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="13" class="abs low">0/13</td>
	</tr>

<tr>
	<td class="file low" data-value="RoomFeedback.js"><a href="RoomFeedback.js.html">RoomFeedback.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="43" class="abs low">0/43</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="18" class="abs low">0/18</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="42" class="abs low">0/42</td>
	</tr>

<tr>
	<td class="file low" data-value="RoomFeedbackResponse.js"><a href="RoomFeedbackResponse.js.html">RoomFeedbackResponse.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="20" class="abs low">0/20</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="19" class="abs low">0/19</td>
	</tr>

<tr>
	<td class="file low" data-value="RoomFeedbackResponseCard.js"><a href="RoomFeedbackResponseCard.js.html">RoomFeedbackResponseCard.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="6" class="abs low">0/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	</tr>

<tr>
	<td class="file low" data-value="RoomFeedbackResponseList.js"><a href="RoomFeedbackResponseList.js.html">RoomFeedbackResponseList.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="102" class="abs low">0/102</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="28" class="abs low">0/28</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="24" class="abs low">0/24</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="94" class="abs low">0/94</td>
	</tr>

<tr>
	<td class="file low" data-value="SingleRoom.js"><a href="SingleRoom.js.html">SingleRoom.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="48" class="abs low">0/48</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="16" class="abs low">0/16</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="45" class="abs low">0/45</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Thu Apr 11 2019 19:58:40 GMT+0300 (East Africa Time)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
<script src="../../block-navigation.js"></script>
</body>
</html>
