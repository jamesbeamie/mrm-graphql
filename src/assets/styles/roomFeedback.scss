@import './variables';

.room-feedback {
  &__control {
    height: 70px;
    text-align: right;
    margin-top: 4%;
    margin-bottom: 4%;
  }

  button {
    border: 1px solid $color-primary;
    border-radius: 2px;
    font-size: 16px;
    color: $color-primary;
    margin-left: 0;
    background-color: map-get($color-white, 000);
    padding: 0 !important;
    height: auto !important;

    @media screen and (max-width: 425px) {
      font-size: 1.05em;
    }
  }

  &__list {
    .table {
      &__row {
        grid-template-columns: 40% 10% 10% 10% 10% 10% 10% !important;
        span {
          display: flex;
        }
      }

      &__headers {
        grid-template-columns: 40% 10% 10% 10% 10% 10% 10% !important;
      }
    }
  }
}

.questions {
  padding: 2%;
  color: $color-primary;
  text-decoration: none;
}
