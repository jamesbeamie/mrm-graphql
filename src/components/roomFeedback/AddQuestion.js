/* eslint-disable import/no-named-as-default */
/* eslint react/no-array-index-key: 0 */
import React, { Component, createRef, Fragment } from 'react';
import moment from 'moment';
import toastr from 'toastr';
import PropTypes from 'prop-types';
import notification from '../../utils/notification';
import Calendar from '../commons/Calendar';
import MrmModal from '../commons/MrmModal';
import { SelectInput as Select } from '../commons';
import '../../assets/styles/addQuestion.scss';
import { addQuestionMutation } from '../helpers/mutationHelpers/questions';
import MultipleInput from '../commons/MultipleInput';


/**
 * Add Question Component
 *
 * @extends React.Component
 *
 * @returns {JSX}
 */
export class AddQuestion extends Component {
  state = {
    checkOptions: [],
    question: '',
    questionTitle: '',
    questionType: 0,
    startDate: moment().format('MMM DD Y'),
    endDate: moment().format('MMM DD Y'),
    startTime: moment().format('HH:MM'),
    endTime: moment()
      .add(1, 'hours')
      .format('HH:MM'),
    calendarOpen: false,
    closeModal: false,
    isLoading: false,
    error: {},
    options: [
      {
        id: '1',
        name: 'Rate',
      },
      {
        id: '2',
        name: 'Check',
      },
      {
        id: '3',
        name: 'Input',
      },
    ],
  };

  /**
   * 1. change isLoading state to it's opposite value
   * i.e true to false or vise verser
   *
   * @returns {void}
   */
  toggleLoading = () => {
    this.setState({
      isLoading: !this.state.isLoading,
    });
  }

  modal = createRef();

  /**
   * It toggles the visibility of the modal
   *
   * @returns {void}
   */
  handleCloseModal = () => {
    this.setState({ closeModal: true });
    this.modal.current.toggleModal();
  };

  /**
   * It toggles the visibility of the calendar modal
   *
   * @returns {void}
   */
  calenderToggle = () => {
    const { calendarOpen } = this.state;
    this.setState({ calendarOpen: !calendarOpen });
  };

  /**
   * It updates the state with the selected start and end date
   *
   * @returns {void}
   */
  sendDateData = (start, end) => {
    this.setState({ startDate: start, endDate: end });
    this.calenderToggle();
  };

  /**
   * It updates the state value of closeModal
   * to false whenever the modal closes
   *
   * @returns {void}
   */
  handleModalStateChange = () => {
    this.state.closeModal && this.setState({ closeModal: false });
  }

  /**
   * It handles the selected choice from the options
   *
   * @param {Object} event
   *
   * @returns {void}
   */
  handleInputChange = (event) => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };

  /**
   * It sets check option state
   *
   * @returns {JSX}
   */
  handleCheckOptions = checkOptions => this.setState({ checkOptions });


  /**
   * This function formats the returned date
   * and time to form a datetime object for
   * both start date and end date
   *
   * @param {date} theDate
   * @param {time} theTime
   *
   * @returns {string}
   */
  formatDateTime = (theDate, theTime) => {
    const theDateValue = (`0${new Date(theDate).getDate()}`).slice(-2);
    const theMonth = (`0${parseInt(new Date(theDate).getMonth() + 1, 10)}`).slice(-2);
    const theYear = new Date(theDate).getFullYear();
    const finalDate = `${theYear}-${theMonth}-${theDateValue}T${theTime}:00`;
    return finalDate;
  };

  /**
   * handles creation of a question
   *
   * @returns {void}
   */
  createQuestion = async () => {
    const {
      question,
      questionTitle,
      questionType,
      startDate,
      endDate,
      startTime,
      endTime,
      options,
      checkOptions,
    } = this.state;
    const valueOfQuestionType = parseInt(questionType, 10) - 1;
    const questionTypeName = options[valueOfQuestionType].name.toLowerCase();
    this.toggleLoading();
    try {
      await addQuestionMutation({
        variables: {
          checkOptions,
          question,
          questionTitle,
          questionType: questionTypeName,
          startDate: this.formatDateTime(startDate, startTime),
          endDate: this.formatDateTime(endDate, endTime),
        },
      });
      notification(
        toastr,
        'success',
        'Question has been added successfully',
      )();
      this.setState({
        question: '',
        questionTitle: '',
        questionType: 0,
        startDate: moment().format('MMM DD Y'),
        endDate: moment().format('MMM DD Y'),
        startTime: moment().format('HH:MM'),
        endTime: moment().format('HH:MM'),
      });
      this.handleCloseModal();
      this.toggleLoading();
      this.props.refetch();
    } catch (err) {
      this.toggleLoading();
      this.handleCloseModal();
      notification(toastr, 'error', err.graphQLErrors[0].message)();
    }
  }

  /**
   * compares the start date with today's date
   * compares the end date with today's date
   * compares the start date with end date
   *
   * @param {date} start
   * @param {date} end
   *
   * @returns {string}
   */
  compareDates = (start, end) => {
    const today = new Date();
    today.setHours(0, 0, 0, 0);
    if (start < today) {
      return this.inputErrors('date', 'Start date cannot be past days');
    } else if (end < today) {
      return this.inputErrors('date', 'End date should be at least a day after today');
    } else if (end <= start) {
      return this.inputErrors('date', 'End date should be at least a day after start date');
    }
    return this.createQuestion();
  }

  /**
   * validates question input before creating a question
   *
   * @param {Object} event
   *
   * @returns {void}
   */
  validateInputFields = () => {
    const {
      question, questionTitle, questionType, startDate, endDate,
    } = this.state;
    if (!questionTitle) {
      return this.inputErrors('questionTitle', 'Please provide the title of the question');
    } else if (!question) {
      return this.inputErrors('question', 'Please provide the feedback question');
    } else if (questionType === 0) {
      return this.inputErrors('type', 'Please select the question type');
    }
    return this.compareDates(new Date(startDate), new Date(endDate));
  }

  /**
   * renders the validation errors for input fields
   *
   * @param {string} errorKey
   * @param {string} message
   *
   * @returns {void}
   */
  inputErrors = (errorKey, message) => {
    this.setState({
      ...this.state.error[errorKey] = message,
    });
  }

  /**
   * It returns the inputs for adding start time and endtime of a question
   * @param {Function} handleInputChange
   *
   * @returns {JSX}
   */
  renderTimeInputs = () => (
    <div className="inputs-inline">
      <div className="inputs">
        <label htmlFor="startTime" className="inputs-title">
          <span>Start</span>
        </label>
        <input
          type="time"
          name="startTime"
          onChange={this.handleInputChange}
          id="timeField"
          value={this.state.startTime}
        />
      </div>
      <div className="inputs">
        <label htmlFor="endTime" className="inputs-title">
          <span>End</span>
        </label>
        <input
          type="time"
          name="endTime"
          onChange={this.handleInputChange}
          id="timeField"
          value={this.state.endTime}
        />
      </div>
    </div>
  );

  /**
   * It returns the input text box for typing the question title
   *
   * @returns {JSX}
   */
  renderQuestionTitleInputBox = () => (
    <div>
      {this.state.error.questionTitle &&
        <span className="span-error">
          <br />{this.state.error.questionTitle}
        </span>}
      <div className={this.state.error.questionTitle ? 'input-error' : 'inputs'}>
        <input
          placeholder="Enter Question Title Here..."
          type="text"
          name="questionTitle"
          onChange={this.handleInputChange}
          value={this.state.questionTitle}
        />
      </div>
    </div>
  );

  /**
   * It returns the input text box for typing the question
   *
   * @returns {JSX}
   */
  renderQuestionInputBox = () => (
    <div>
      {this.state.error.question &&
        <span className="span-error">
          <br />{this.state.error.question}
        </span>}
      <div className={this.state.error.question ? 'input-error' : 'inputs'}>
        <input
          placeholder="Enter Question here..."
          type="text"
          name="question"
          onChange={this.handleInputChange}
          value={this.state.question}
        />
      </div>
    </div>
  );

  /**
   * It renders the calendar component
   *
   * @returns {JSX}
   */
  renderCalendarComponent = () => (
    <div className="calendar-container">
      {this.state.error.date &&
        <span className="span-error">
          <br />{this.state.error.date}
        </span>}
      <Calendar
        disabledDateRange="past"
        sendData={this.sendDateData}
        classProp={this.state.error.date ? 'calendarIconBtn-error' : 'calendarIconBtn'}
      />
    </div>
  );


  /**
   * It renders the dropdown for
   * selecting question type
   *
   * @returns {JSX}
   */
  rendersQuestionType = () => (
    <div>
      {this.state.error.type &&
        <span className="span-error">
          <br />{this.state.error.type}
        </span>}
      <Select
        labelText=""
        name="questionType"
        id={this.state.error.type ? 'selectType-error' : 'selectType'}
        value={this.state.questionType}
        onChange={this.handleInputChange}
        wrapperClassName="input-wrapper"
        placeholder="Select Type"
        options={this.state.options}
      />
    </div>
  );


  /**
   * It shows all the input fields required for creating a question
   *
   * @returns {JSX}
   */
  renderQuestionValues = () => (
    <div className="question-form">
      <span className="question-form__sections">Date</span>
      {this.renderCalendarComponent()}
      <span className="question-form__sections">Time</span>
      {this.renderTimeInputs()}
      <span className="question-form__sections">Question Title</span>
      {this.renderQuestionTitleInputBox()}
      <span className="question-form__sections">Question</span>
      {this.renderQuestionInputBox()}
      <span className="question-form__sections">Question Type</span>
      {this.rendersQuestionType()}
      {
        this.state.questionType === '2'
        && (
          <Fragment>
            <span className="question-form__sections">Check Options</span>
            <MultipleInput submit={this.handleCheckOptions} placeholder="Enter option and press 'Enter' or ','" />
          </Fragment>
        )
      }
    </div>
  );

  render() {
    return (
      <MrmModal
        ref={this.modal}
        title="ADD QUESTION"
        buttonText="Add Question"
        modalButtonClassName="add-question-button"
        modalContent={this.renderQuestionValues()}
        cancelButtonText="CANCEL"
        handleSubmit={this.validateInputFields}
        isLoading={this.state.isLoading}
        iconButtonClass="modal-trigger"
      />
    );
  }
}

AddQuestion.propTypes = {
  refetch: PropTypes.func.isRequired,
};

export default AddQuestion;
