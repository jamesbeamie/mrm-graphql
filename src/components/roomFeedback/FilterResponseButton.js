import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { filterYellowIcon } from '../../utils/images/images';

class FilterResponseButton extends Component {
  state = {

  }
  render() {
    const { toggleFilterModal } = this.props;
    return (
      <div>
        <button
          className="btn-secondary response-filter-button"
          onClick={toggleFilterModal}
        >
          <div>
            <img
              src={filterYellowIcon}
              alt="Filter"
            />
            <span>Filter</span>
          </div>
        </button>
      </div>
    );
  }
}

FilterResponseButton.propTypes = {
  toggleFilterModal: PropTypes.func.isRequired,
};

export default FilterResponseButton;
