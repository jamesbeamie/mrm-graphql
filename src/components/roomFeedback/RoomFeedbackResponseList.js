import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import RoomFeedbackResponse from './RoomFeedbackResponse';
import { greenStarIcon, greyStarIcon } from '../../utils/images/images';
import '../../../src/assets/styles/roomFeedbackResponseList.scss';
import '../../../src/assets/styles/feedbackContainer.scss';
import RoomFeedbackCard from './RoomFeedbackResponseCard';
import SingleRoom from './SingleRoom';
import Pagination from '../../../src/components/commons/Pagination';
import Overlay from '../commons/Overlay';
import paginate from '../helpers/FrontendPagination';

/**
 * Loops through the room responses
 * and calculates the total number of missing
 * items
 *
 * @param {array} roomResponses
 *
 * @return {integer}
 */
export const roomCleanlinessRating = (rating) => {
  let rated = rating;
  const data = [];
  while (rated > 0) {
    data.push(<img key={rated} src={greenStarIcon} alt="" />);
    rated -= 1;
  }
  let unRated = 5 - rating;
  while (unRated > 0) {
    data.push(<img key={5 + unRated} src={greyStarIcon} alt="" />);
    unRated -= 1;
  }
  return data;
};

/**
 * Loops through the room responses
 * and calculates the total cleanliness
 * rating and grade
 *
 * @param {array} roomResponses
 *
 * @return {integer, string}
 */
export const totalCleanlinessRating = (roomResponses = []) => {
  let totalRating = 0;
  const gradeList = ['Not Rated', 'Poor', 'Fair', 'Good', 'Very Good', 'Excellent'];
  const responsesWithRating = roomResponses.filter(response => response.response.__typename === 'Rate');
  responsesWithRating.forEach((singleResponse) => { totalRating += singleResponse.response.rate; });
  let averageRating = Math.round(totalRating / (responsesWithRating.length));
  if (Number.isNaN(averageRating)) {
    averageRating = 0;
  }
  const grade = gradeList[(averageRating)];
  return { averageRating, grade };
};

/**
 * Loops through the room responses
 * and calculates the total number of missing
 * items
 *
 * @param {array} roomResponses
 *
 * @return {integer}
 */
export const totalMissingItemsCount = (roomResponses = []) => {
  let missingItemsList = [];
  const responsesWithMissingItems = roomResponses.filter(response => response.response.__typename === 'MissingItems');
  responsesWithMissingItems.forEach((response) => {
    missingItemsList = [...new Set(response.response.missingItems)];
  });
  return missingItemsList.length;
};

/**
 * Calculates the total number of room resources
 *
 * @param {array} roomResources
 *
 * @return {integer}
 */
export const totalRoomResources = (roomResources = []) => {
  const totalRoomResourcesNumber = roomResources.reduce(
    (accumulator, currentValue) => accumulator + currentValue.room[0].quantity,
    0,
  );
  return totalRoomResourcesNumber;
};

/**
 * Component for RoomFeedbackResponseList
 *
 * @extends React.Component
 *
 * @returns {JSX}
 */
export class RoomFeedbackResponseList extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      visible: false,
      roomId: null,
      responseIds: [],
      currentPage: 1,
      isFetching: false,
    };
  }

  /**
   * Sets component state after comparing roomId state with id argument
   *
   * @param {Object} event
   * @param {string} id
   */
  showModal = (e, data = null) => {
    e.preventDefault();
    if (data) {
      const { roomId, responseIds } = data;
      this.setState({ visible: true, roomId, responseIds });
    } else {
      this.setState({ visible: false, roomId: null, responseIds: [] });
    }
  };

  /**
   * Queries resources list
   *
   * @param {Integer} perPage
   * @param {Integer} page
   *
   * @returns {void}
   */
  handleData = (perPageData, currentPageData) => this.setState({
    perPage: perPageData,
    currentPage: currentPageData,
  });

  /**
   * Renders a card
   *
   * @returns {JSX}
   */
  renderCard = (feedbackData) => {
    const { totalResponses, roomsWithMissingItems, totalAverageRating } = feedbackData;
    const cardsContent = [
      {
        title: 'Total Responses',
        value: totalResponses,
      },
      {
        title: 'Rooms With Missing Items',
        value: roomsWithMissingItems,
      },
      {
        title: 'Average Ratings',
        value: totalAverageRating,
      },
    ];
    return cardsContent.map((cardData, index) => (
      /* eslint-disable-next-line react/no-array-index-key */
      <RoomFeedbackCard {...cardData} key={index} />
    ));
  };

  render() {
    const {
      roomId, isFetching, currentPage, perPage, responseIds,
    } = this.state;
    const { feedback, loading } = this.props;

    const feedbackData = loading
      ? {
        roomsResponses: [{
          roomId: 1,
          roomName: 'Sample',
          response: [],
          totalResponses: 1,
          totalRoomResources: 1,
        }],
      } : {
        ...feedback,
        ...paginate(feedback.roomsResponses, { currentPage, perPage }, { dataName: 'roomsResponses' }),
      };

    const displayPaginator = !isFetching
      && !(feedbackData.roomsResponses.length < 5 && currentPage === 1);

    if (loading || (feedbackData.roomsResponses).length > 0) {
      return (
        <Fragment>
          <div className="feedback-container overlay-container">
            {loading && <Overlay />}
            <SingleRoom
              roomId={roomId}
              visible={this.state.visible}
              showModal={this.showModal}
              responseIds={responseIds}
              totalCleanlinessRating={totalCleanlinessRating}
              roomCleanlinessRating={roomCleanlinessRating}
              totalMissingItemsCount={totalMissingItemsCount}
              totalRoomResources={totalRoomResources}
            />
            <div className="feedback-card ">{this.renderCard(feedbackData)}</div>
            <div className="feedback-reponses-list">
              {isFetching ? <Overlay /> : null}
              <header className="room-feedback-list-container">
                <span>Meeting Room</span>
                <span>Responses</span>
                <span>Cleanliness Rating</span>
                <span>Missing Items</span>
                <span>Suggestion on how to improve</span>
              </header>
              {(feedbackData.roomsResponses).map(singleFeedback => (
                <RoomFeedbackResponse
                  activeRoomId={roomId}
                  roomFeedbackResponse={singleFeedback}
                  key={singleFeedback.roomId}
                  viewSingleFeed={this.showModal}
                  totalCleanlinessRating={totalCleanlinessRating}
                  roomCleanlinessRating={roomCleanlinessRating}
                  totalMissingItemsCount={totalMissingItemsCount}
                  totalRoomResources={totalRoomResources}
                />
              ))}
            </div>
            {displayPaginator && <Pagination
              totalPages={feedbackData.pages}
              hasNext={feedbackData.hasNext}
              hasPrevious={feedbackData.hasPrevious}
              handleData={this.handleData}
              currentPage={currentPage}
              isFetching={isFetching}
              perPage={perPage ? Number(perPage) : 5}
            />}
          </div>
        </Fragment>
      );
    }
    return (
      <div className="item-list-empty">
        <div className="message">There are no rooms with responses</div>
      </div>
    );
  }
}

RoomFeedbackResponseList.propTypes = {
  loading: PropTypes.bool.isRequired,
  feedback: PropTypes.shape({}).isRequired,
};

export default RoomFeedbackResponseList;
