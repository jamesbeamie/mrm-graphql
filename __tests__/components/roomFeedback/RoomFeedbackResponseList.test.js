import React from 'react';
import { shallow } from 'enzyme';
import { RoomFeedbackResponseList, roomCleanlinessRating } from '../../../src/components/roomFeedback/RoomFeedbackResponseList';

const responseListProps = {
  loading: false,
  feedback: {
    roomsResponses: [
      {
        roomId: 1,
        roomName: 'Olkaria',
        totalResponses: 25,
        totalRoomResources: 5,
        response: [
          {
            id: 289,
            createdDate: '2019-06-25T14:45:26.577611',
            resolved: false,
            response: {
              __typename: 'Rate',
              rate: 5,
            },
          },
          {
            id: 290,
            createdDate: '2019-06-25T14:45:27.942964',
            resolved: false,
            response: {
              __typename: 'MissingItems',
              missingItems: { id: 'test', name: 'test name' },
            },
          },
          {
            id: 291,
            createdDate: '2019-06-25T14:46:06.718295',
            resolved: false,
            response: {
              __typename: 'SelectedOptions',
              options: [
                'handywork',
              ],
            },
          },
        ],
      },
    ],
  },
};


const propsWithoutResponses = {
  loading: false,
  feedback: {
    roomsResponses: [],
  },
};

describe('Room feedback list', () => {
  it('should render correctly', () => {
    const wrapper = shallow(
      <RoomFeedbackResponseList {...responseListProps} checkData={jest.fn()} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a spinner when loading prop is true', () => {
    const loadingProps = {
      ...propsWithoutResponses,
      loading: true,
    };
    const wrapper = shallow(
      <RoomFeedbackResponseList {...loadingProps} checkData={jest.fn()} />);
    expect(wrapper.find('Overlay')).toHaveLength(1);
  });

  it('should render correctly when no room has an error response', () => {
    const emptyComponent = shallow(
      <RoomFeedbackResponseList {...propsWithoutResponses} checkData={jest.fn()} />);
    expect(emptyComponent.find('.item-list-empty')).toHaveLength(1);
  });

  it('returns the correct star rating markup', () => {
    const result = roomCleanlinessRating(1);
    expect(result.length).toEqual(5);
  });
});
