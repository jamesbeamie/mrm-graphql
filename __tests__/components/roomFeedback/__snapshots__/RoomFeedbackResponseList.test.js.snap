// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Room feedback list should render correctly 1`] = `
<Fragment>
  <div
    className="feedback-container overlay-container"
  >
    <SingleRoomFeedBack
      responseIds={Array []}
      roomCleanlinessRating={[Function]}
      roomId={null}
      showModal={[Function]}
      totalCleanlinessRating={[Function]}
      totalMissingItemsCount={[Function]}
      totalRoomResources={[Function]}
      visible={false}
    />
    <div
      className="feedback-card "
    >
      <RoomFeedbackCard
        key="0"
        title="Total Responses"
        value={0}
      />
      <RoomFeedbackCard
        key="1"
        title="Rooms With Missing Items"
        value={0}
      />
      <RoomFeedbackCard
        key="2"
        title="Average Ratings"
        value={0}
      />
    </div>
    <div
      className="feedback-reponses-list"
    >
      <header
        className="room-feedback-list-container"
      >
        <span>
          Meeting Room
        </span>
        <span>
          Responses
        </span>
        <span>
          Cleanliness Rating
        </span>
        <span>
          Missing Items
        </span>
        <span>
          Suggestion on how to improve
        </span>
      </header>
      <RoomFeedbackResponse
        activeRoomId={null}
        key="1"
        roomCleanlinessRating={[Function]}
        roomFeedbackResponse={
          Object {
            "response": Array [
              Object {
                "createdDate": "2019-06-25T14:45:26.577611",
                "id": 289,
                "resolved": false,
                "response": Object {
                  "__typename": "Rate",
                  "rate": 5,
                },
              },
              Object {
                "createdDate": "2019-06-25T14:45:27.942964",
                "id": 290,
                "resolved": false,
                "response": Object {
                  "__typename": "MissingItems",
                  "missingItems": Object {
                    "id": "test",
                    "name": "test name",
                  },
                },
              },
              Object {
                "createdDate": "2019-06-25T14:46:06.718295",
                "id": 291,
                "resolved": false,
                "response": Object {
                  "__typename": "SelectedOptions",
                  "options": Array [
                    "handywork",
                  ],
                },
              },
            ],
            "roomId": 1,
            "roomName": "Olkaria",
            "totalResponses": 25,
            "totalRoomResources": 5,
          }
        }
        totalCleanlinessRating={[Function]}
        totalMissingItemsCount={[Function]}
        totalRoomResources={[Function]}
        viewSingleFeed={[Function]}
      />
    </div>
  </div>
</Fragment>
`;
